spec/ddtrace/utils_spec.rb

```diff
@@ -0,0 +1,141 @@
+require 'concurrent-ruby'
+
+RSpec.describe Datadog::Utils do
+  describe '.next_id' do
+    subject(:next_id) { described_class.next_id }
+
+    it 'returns a positive integer smaller than 2**62' do
+      is_expected.to be_a(Integer)
+      is_expected.to be_between(1, 2**62 - 1)
+    end
+
+    it 'fits in a CRuby VALUE slot', if: Datadog::Core::Environment::ObjectSpace.estimate_bytesize_supported? do
+      expect(Datadog::Core::Environment::ObjectSpace.estimate_bytesize(next_id)).to eq(0)
+    end
+
+    it 'returns unique numbers on successive calls' do
+      # This is the same assumption of non-collision we have in our production systems.
+
+      ids = Array.new(100) { described_class.next_id }.uniq
+      expect(ids).to have(100).items
+    end
+
+    it 'returns unique numbers in a threaded environment' do
+      barrier = Concurrent::CyclicBarrier.new(100)
+      threads = Array.new(100) do
+        Thread.new do
+          barrier.wait
+          described_class.next_id
+        end
+      end
+
+      ids = threads.map(&:value).uniq
+      expect(ids).to have(100).items
+    end
```

Comment:
I'm not entirely convinced that this test is useful, because it seems to me that we're testing that the Ruby RNG implementation is thread-safe.

The thing is: I think this is a poor test for thread safety. I think if there are thread safety issues, I'm not convinced this test will expose them.

Like the test above, if the test is just a check to see if we're using thread-local randoms that are initialized all with the same seed (what are the odds?), then checking with just two calls should get us the same validation but again... I'm not entirely convinced that such a black-box test provides actual value.

Reply:
If we've moved this implementation to Ruby-land, and goofed the random number generation by being possible to have two simultaneous calls to `next_id` return the same number, do you think this test would be useful?

If not, do you know what kind of test would be in this case? It's always tricky to "test" this kind of issue.

Reply:
> If we've moved the next_id implementation to Ruby-land, and goofed the random number generation by being possible to have two simultaneous calls to next_id return the same number (read/write kind of race condition), do you think this test would be useful?

I don't. For race conditions, we probably want to have a stress test that runs for a reasonable amount of time in a loop and tries to exercise the case hard; doing it once in this way is a bit like playing in the lottery: sure, the chances of you winning are not zero, but do we really expect to win?

> If not, do you know what kind of test would be in this case? It's always tricky to "test" this kind of issue.

To be honest, I think that we should instead treat the Ruby implementation as a black box that we assume correct, and assert that the code calls the ruby implementation exactly as we expect. Such a test would also fail if one were to change the "next_id" implementation to some Ruby code, because we'd not be following the spec of calling the ruby implementation in the same way.

Then, if we really are concerned, we could have a separate stress test for the Ruby API.